ducktape/cluster/vagrant.py

```diff
@@ -24,10 +28,53 @@ class VagrantCluster(JsonCluster):
     advertised, this assumes that the Vagrant VM's name is a routeable hostname on all the hosts.
     """
 
-    def __init__(self):
-        hostname, ssh_hostname, username, flags = None, None, None, ""
-        nodes = []
+    def __init__(self, *args, **kwargs):
+        is_read_from_file = False
 
+        cluster_file = kwargs.get("cluster_file")
+        if cluster_file is not None:
+            try:
+                cluster_json = json.load(open(os.path.abspath(cluster_file)))
+                is_read_from_file = True
+            except IOError:
+                # It is OK if file is not found. Call vagrant ssh-info to read the cluster info.
+                pass
+
+        if not is_read_from_file:
+            cluster_json = {
+                "nodes": self._get_nodes_from_vagrant()
+            }
+
+        super(VagrantCluster, self).__init__(cluster_json)
+
+        # Load externally_routable_ip for each account
+        if is_read_from_file:
+            externally_routable_ip = cluster_json["externally_routable_ip"]
```

Comment:
This bit was confusing to me at first, but I think I understand now what's going on. 

This actually makes the schema of the Vagrant cluster json different from that of a corresponding JsonCluster, and I think this is not quite what we want. See if this makes sense to you:

If I cache cluster info and examine the `RemotAccount` objects created with a call like this

`ducktape test/path --cluster-file cluster.json`

I would expect those `RemoteAccount` objects to be identical to `RemoteAccount` objects created if I then instantiated the cluster this way:

`ducktape test/path --cluster ducktape.cluster.json.JsonCluster --cluster-info cluster.json`

This isn't true because in the first case, the `RemoteAccount` objects will have a non-None `externally_routable_ip` field, but in the second case, `externally_routable_ip` field will be None.

---

In other words, as long as `cluster.json` exists, it's conceptually simpler if these two commands are completely interchangeable:
`ducktape test/path --cluster-file cluster.json`
`ducktape test/path --cluster ducktape.cluster.json.JsonCluster --cluster-info cluster.json`

---

I think the json written to file should look like this: 

```
{
  "nodes": [{"ssh_hostname": "127.0.0.1",
                   "externally_routable_ip": "192.68.asfdasdfasd"
                   "hostname": "ducktape1", 
                   "user": "vagrant", 
                   "ssh_args": "-o 'HostName 127.0.0.1' -o 'Port 2222' -o 'UserKnownHostsFile /dev/null' -o 'StrictHostKeyChecking no' -o 'PasswordAuthentication no' -o 'IdentityFile /Users/geoffreyanderson/Google_Drive/Confluent_code/ducktape-dong/.vagrant/machines/ducktape1/virtualbox/private_key' -o 'IdentitiesOnly yes' -o 'LogLevel FATAL' "
}]
}
```

and `JsonCluster` can instantiate its list of `RemoteAccount` objects with the additional `externally_routable_ip` field.

Does this seem reasonable?